### ISSUE TITLE
GPU accelerated method producing different result from non-accelerated method

### ISSUE THREAD

**Issue by lriordanchen** [tomopy/tomopy#425]
Hello,

I'm having trouble implementing the GPU support.

When I run the reconstruction without GPU support, it works as expected, but when I add in the line to run it on the GPU, the reconstruction contains artifacts consistent with using the incorrect rotation center in the reconstruction algorithm.

```python
import tomopy

# import data, set theta, find rotation center using tomopy.find_center_pc

params = {'algorithm':'sirt', 'num_iter':10}
recon = zeros(shape=(num_channels, image_height, image_width, image_width))
for ix in range(num_channels):
    print('Reconstructing channel: ', ix+1)
    recon_channel = tomopy.recon(align_proj[ix], theta, center=rot_center[ix], 
                                 **params,
                                 accelerated=True, device="gpu" 
                                 )
    recon[ix] = recon_channel
```
So when I comment out the line <accelerated=True, device="gpu" >, the reconstruction looks reasonable but when I run it as above, I see the characteristic loops and double edges as if the rotation center is incorrect.  What could be the reason for this, and how can I get this to reconstruct correctly when I run on the GPU?

I'm running on:
 - OS: Windows 10
 - TomoPy 1.5.0

**Comment by jrmadsen** [comment]
Hi @lriordanchen, the GPU algorithm sets up the problem differently than the CPU algorithms -- instead of keeping the ROI stationary and calculating the chord through the pixel at a given projection angle, it rotates the entire ROI to be perpendicular with the projection angle and interpolates the new pixels. at this point, the algorithms proceeds essentially the same as the CPU algorithms except all the weights are 1. Then the ROI is back rotated and interpolated to the original coordinates. 

So TL;DR -- you need to pad your sinogram with ones or zeros so a rotation of 45 degrees doesn't result in a data loss. 

For the built-in phantoms, here is an example of the padding being applied: https://github.com/tomopy/tomopy/blob/cfeb9cab03a85063e43ef307a990497e98ccd17a/benchmarking/phantom.py#L36 

@dgursoy or @carterbox can probably give you more exact/accurate instructions on how to do this though.

**Comment by carterbox** [comment]
@jrmadsen, you're describing a different problem than what @lriordanchen is reporting. The problem is that in an experiment, the center of rotation for the object isn't always perfectly aligned with the field of view. This is why we have the `center` parameter. `sirt_cuda` takes a parameter `const float* center` which is an array of x (horizontal) coordinates. This is the location of the "true" center of rotation in each slice. However, this parameter seems to be unused by `sirt_cuda`.

https://github.com/tomopy/tomopy/blob/cfeb9cab03a85063e43ef307a990497e98ccd17a/source/gpu/sirt.cu#L162-L247

Right now it looks like we are assuming that the center of rotation is located at `(dx/2, dy/2)`. 

https://github.com/tomopy/tomopy/blob/cfeb9cab03a85063e43ef307a990497e98ccd17a/source/gpu/utils.cu#L244-L245

However, the old sirt uses the `center` parameter so the center of rotation is `(center, dy/2)`.

I haven't run my own tests to confirm this is the case, but if @lriordanchen posts a photo of one of their reconstructions, we can easily confirm that this is the center issue and not a padding issue.

@lriordanchen, as a temporary measure, you can adjust the rotation centers of the sinograms before handing them off to SIRT by shifting them along the x-axis. However, the reason why the `center` parameter exists is probably to avoid interpolation errors from shifting the sinogram.

**Comment by carterbox** [comment]
This is both a bug and feature request. At the very least, there should be a warning that `sirt_cuda` does not have feature parity with `sirt`, but it is also requesting that a feature is implemented.

**Comment by jrmadsen** [comment]
Ah ok. Thanks for correcting this. Yes it should not be difficult to adapt the affine transform in [utils.cu](https://github.com/tomopy/tomopy/blob/cfeb9cab03a85063e43ef307a990497e98ccd17a/source/gpu/utils.cu#L244-L245) to handle this.

**Comment by lriordanchen** [comment]
Thanks @jrmadsen and @carterbox!  Here is an example of what I was seeing:

Below I have an slice generated by running _without_ the gpu:
![image_no_gpu](https://user-images.githubusercontent.com/36885244/60593301-bd6fe300-9da2-11e9-94d8-8f7cdafcde93.jpg)

and here I have the same image generated by running _with_ the gpu:
![image_gpu](https://user-images.githubusercontent.com/36885244/60593340-cf518600-9da2-11e9-86a3-1c2e22d1f1ca.jpg)

Do you agree that it is a problem with the rotation center?  Also, I found the same thing happened with mlem.

Adding on to the feature request in issue #422, would it be possible to have the padding @jrmadsen talked about automatically taken care of after the code has determined not to fallback to the cpu?  That way the result, whether arrived at by using the gpu as asked for or by using the cpu as fallback, is calculated correctly.

Thanks again for all your help.

**Comment by dgursoy** [comment]
Thanks for bringing this up.

Currently the GPU implementation assumes that the center is at the middle of the detector's field of view no matter what you provide with the ```center``` argument. For a quick fix, one can shift the data by padding or interpolation such that it is centered correctly. 

But in the long run we need to fix it in TomoPy. This requires implementation of rotation around an arbitrary point and deserves a new pull request. First, we should generate a warning for users that ```center``` is currently a dummy argument. 

I'll post it here once the PR is open.

**Comment by jdgelb** [comment]
I've noticed that this issue also seems to apply when using Astra for SIRT (haven't yet tried any of the other Astra recon modules). Passing in the `center` argument seems to have no effect on the reconstruction (i.e., it assumes the center is the middle of the image even though I have specified the center shift). This issue does not apply to TomoPy's own Gridrec. 

Here's my recon call for SIRT using Astra:
```
options = {'proj_type': 'cuda', 'method': 'SIRT_CUDA', 'num_iter': 100}
recon = tomopy.recon(proj, theta, center=484, algorithm=tomopy.astra, options=options, ncore=1)
```
And here is my call for Gridrec:
`recon = tomopy.recon(proj, theta, center=484, algorithm='gridrec', filter_name='shepp', ncore=16)`
Below is the output of each... the Gridrec results have clearly been center-shifted, while the Astra ones have not. 
![image](https://user-images.githubusercontent.com/37162010/61407856-967dda80-a893-11e9-82be-a30533311909.png)

**Comment by decarlof** [comment]
@jdgelb as mentioned by @dgursoy you can just shift with:

```
options = {'proj_type': 'cuda', 'method': 'SIRT_CUDA', 'num_iter': 100}
shift = 28
proj = np.roll(proj, shift, axis=2)
recon = tomopy.recon(proj, theta, algorithm=tomopy.astra, options=options, ncore=1)
```
assuming you had a 1024 detector (512 - 484) = 28

**Comment by canismarko** [comment]
I think the @jdgelb astra issue is different from the tomopy GPU. I was getting ready to open a new issue but since it's related I'll post it here.

In the astra implementation, tomopy accounts for the *center* parameter by [setting the *ExtraDetectorOffset* option](https://github.com/tomopy/tomopy/blob/139f4e82beea6b6dbdf54b9f531215d42dc59b5b/source/tomopy/recon/wrappers.py#L192) on the projection geometry object.

However, this doesn't actually change the reconstruction in astra, as demonstrated in [this MWE](https://gist.github.com/canismarko/be0b5d0f4766eb0b603ca6b599271ea5). The example is a modification of [sample s003](https://github.com/astra-toolbox/astra-toolbox/blob/master/samples/python/s003_gpu_reconstruction.py), with the addition of the extra detector offset on line 34. If I run the example, the reconstruction is the same no matter what value I put in for *center*.

I'm not sure on whether astra or tomopy has the issue, though.

**Comment by Dinghy666** [comment]
I had this problem as well, so is this problem solved now？
